### PR TITLE
RFC: separate object lookup error for promisor packs

### PR DESCRIPTION
**Pull request by rcoup**
There's currently no support in libgit2 for partial clones (#5564). This attempts to start support them by distinguishing object-lookup faults between `ENOTFOUND` (the requested oid is bad or the repo is corrupt) from a new `EMISSING` error (indicating the oid is referenced from a promisor pack, and therefore _should_ be available remotely).

1. Detect a promisor packfile by looking for a `.packfile` sidecar and track it in an attribute of `git_pack_file`
2. When an object is referenced from a promisor packfile (eg. tree_entry → blob), raise EMISSING instead of ENOTFOUND if the object isn't present anywhere in the ODB.

To be clear, there's no support for _fetching_ missing objects from a promisor remote.

The code doesn't actually do (2) properly when there are multiple ODB backends in place, but hopefully it conveys the basic idea.

Conceptually, EMISSING could also be used for things like shallow clones where parent commits aren't available locally, etc.

### CODE REVIEW

**Comment by olsen232** [comment]
Hi github maintainers, I'm willing to do some work to help improve libgit2's partial clone support. What can I do to help this along?

File: src/pack.c [hunk 1]
```diff
@@ -1190,15 +1194,30 @@ int git_packfile_alloc(struct git_pack_file **pack_out, const char *path)
...
+		root_len = path_len - strlen(".idx");
 		if (!git_disable_pack_keep_file_checks) {
 			memcpy(p->pack_name + root_len, ".keep", sizeof(".keep"));
 			if (git_path_exists(p->pack_name) == true)
 				p->pack_keep = 1;
 		}
 
 		memcpy(p->pack_name + root_len, ".pack", sizeof(".pack"));
+	} else {
+		root_len = path_len - strlen(".pack");
```
Comment (ethomson) [inline src/pack.c:1206]: In [line 1181](https://github.com/libgit2/libgit2/pull/5776/files#diff-226580d1635752a6f3386d44293260fbd4a33f2f405ba0b91fbd35c3d96315f9R1181) we ensure that `path_len` is at least `strlen(".idx")`.  If it were `strlen(".idx")`, here we would set `root_len` to `(size_t)-1`.

I think that it's unlikely that there's a call path that would call us with just `.idx` as the path but some sort of guard seems like it should be in order.

**Review by ethomson — COMMENTED** [review]
Sorry about the delay, @rcoup; I haven't had a lot of bandwidth in 2020.  I think that the direction that this is going is largely correct.

> The code doesn't actually do (2) properly when there are multiple ODB backends in place, but hopefully it conveys the basic idea.

Is the problem here that we're short circuiting the lookup on the first promisor pack, and a subsequent odb backend may have the object?  (In other words, we erroneously report `EMISSING` even though we have the object?)  Or is there a deeper underlying problem that I'm not seeing yet?

> Conceptually, EMISSING could also be used for things like shallow clones where parent commits aren't available locally, etc.

I wonder if that's what we would want to do.  I agree that they're similar (object exists on the remote and doesn't exist locally) but I wonder if they should be distinct error codes so that a caller can distinguish them more easily.  🤔

**Comment by rcoup** [comment]
> Is the problem here that we're short circuiting the lookup on the first promisor pack, and a subsequent odb backend may have the object? (In other words, we erroneously report EMISSING even though we have the object?) 

Yes. Presumably there might be some other complexities wrt multiple ODB backends.

>>Conceptually, EMISSING could also be used for things like shallow clones where parent commits aren't available locally, etc.
> 
> I wonder if that's what we would want to do. I agree that they're similar (object exists on the remote and doesn't exist locally) but I wonder if they should be distinct error codes so that a caller can distinguish them more easily. 🤔

Yes, maybe `EPROMISED` & `EGRAFT` (or `ESHALLOW_GRAFT`?) to make it clearer?

**Comment by olsen232** [comment]
FYI I picked up this PR and started to address the review comments - however I ran into a major obstacle - `git_object_lookup` doesn't actually have the necessary context to know if an object is non-existent, missing-without-good-reason (ie corrupt), or promised. In order to say that an object is promised, we should ensure the following:

1. The object is not in the ODB.
2. The object is referenced by an object that *is* in the ODB.
3. The object that *is* in the ODB is in a packfile, and, that packfile is a promisor packfile.
Since `git_object_lookup` doesn't have any context except the OID that we're looking for, we can't make the distinction at this layer without changing the API.

This PR doesn't currently do those things and lacks the context to do so without changing the libgit2 API. I think this PR promotes ENOTFOUND to EPROMISED basically every time if there is a promisor packfile involved, which is too eager (unless we define the error code as meaning "this object might be promised", as opposed to "is promised"). I think this PR is probably best simply closed - there are two other PRs that I will try to write instead:

- Separate error code eg EMISSING for when an object lookup by path fails due to missing object (not due to an invalid path). For instance, if I run `git_tree_entry_bypath(&entry, root, "ab/cd/ef") and "ab/cd/ef" is not a valid path - perhaps ab has no child cd - then I get ENOTFOUND, as I should. But if I run `git_tree_entry_bypath(&entry, root, "ab/cd/ef")` and "ab/cd/ef" is valid, as far as can be tested, but an object is missing, I should get this new error instead. (Eg if ab does have a child cd, but that child is nowhere to be found.

- Separate error code eg EPROMISED for when an object is EMISSING (as in the last PR) *AND* the packfile where we would expect to find the object is a promisor packfile. I would salvage some of this PR to check if the promisor file exists.

Limitations - I can only add these codes to the path API eg `git_tree_entry_bypath`, `git_tree_entry_byname`, `git_object_lookup_bypath`. git itself also has this limitation eg if you run `git cat-file $SOME_OBJECT_ID` it doesn't have enough context to know if the object is non-existent or promised, so it has to search all promisor backends for it.

**Comment by rcoup** [comment]
Closed in favour of #5993 and follow-ups

**Pull request closed without merging**